pkg/scheduler/queue/tree_queue.go

```diff
@@ -0,0 +1,282 @@
+//
+// TreeQueue internally maintains round-robin fair queuing across all of its queue dimensions.
+// Each queuing dimension is modeled as a node in the tree, internally reachable through a QueuePath.
+//
+// The QueuePath is an ordered array of strings describing the path through the tree to the node,
+// which contains the FIFO local queue of all items enqueued for that queuing dimension.
+//
+// When dequeuing from a given node, the node will round-robin equally between dequeuing directly
+// from its own local queue and dequeuing recursively from its list of child TreeQueues.
+// No queue at a given level of the tree is dequeued from consecutively unless all others
+// at the same level of the tree are empty down to the leaf node.
+type TreeQueue struct {
+	// name of the tree node will be set to its segment of the queue path
+	name                   string
+	maxQueueLen            int
+	localQueue             *list.List
+	currentChildQueueIndex int
+	childQueueOrder        []string
+	childQueueMap          map[string]*TreeQueue
+}
+
+func NewTreeQueue(name string, maxQueueLen int) *TreeQueue {
+	return &TreeQueue{
+		name:                   name,
+		maxQueueLen:            maxQueueLen,
+		localQueue:             list.New(),
+		currentChildQueueIndex: localQueueIndex,
+		childQueueMap:          map[string]*TreeQueue{},
+		childQueueOrder:        nil,
+	}
+}
+
+func (q *TreeQueue) IsEmpty() bool {
+	// avoid recursion to make this a cheap operation
+	//
+	// Because we dereference empty child nodes during dequeuing,
+	// we assume that emptiness means there are no child nodes
+	// and nothing in this tree node's local queue.
+	//
+	// In reality a package member could attach empty child queues with getOrAddNode
+	// in order to get a functionally-empty tree that would report false for IsEmpty.
+	// We assume this does not occur or is not relevant during normal operation.
+	return q.localQueue.Len() == 0 && len(q.childQueueMap) == 0
+}
+
+func (q *TreeQueue) NodeCount() int {
+	count := 1 // count self
+	for _, childQueue := range q.childQueueMap {
+		count += childQueue.NodeCount()
+	}
+	return count
+}
+
+func (q *TreeQueue) ItemCount() int {
+	count := q.localQueue.Len() // count self
+	for _, childQueue := range q.childQueueMap {
+		count += childQueue.ItemCount()
+	}
+	return count
+}
+
+// EnqueueBackByPath enqueues an item in the back of the local queue of the node
+// located at a given path through the tree; nodes for the path are created as needed.
+//
+// childPath must be relative to the receiver node; providing a QueuePath beginning with
+// the receiver/parent node name will create a child node of the same name as the parent.
+func (q *TreeQueue) EnqueueBackByPath(childPath QueuePath, v any) error {
+	childQueue, err := q.getOrAddNode(childPath)
+	if err != nil {
+		return err
+	}
+	if childQueue.localQueue.Len()+1 > childQueue.maxQueueLen {
```

Comment:
I know this PR does not introduce the queuing by component, but I'm wondering if this is the right logic here - don't we want to continue limiting the total size of a tenant's queue, not the size of a component's queue?

(If the answer to this is "it's coming in a subsequent PR", that's perfectly fine 🙂)